ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/WebSocketRequestHandler.java

```diff
@@ -230,4 +254,60 @@ private JsonRpcResponse errorResponse(final Object id, final JsonRpcError error)
   private boolean isNonEmptyResponses(final JsonRpcResponse result) {
     return result.getType() != JsonRpcResponseType.NONE;
   }
+
+  static class JsonResponseStreamer extends OutputStream {
+    private final ServerWebSocket response;
+    private final Semaphore paused = new Semaphore(0);
+    private final byte[] singleByteBuf = new byte[1];
+    private boolean firstFrame = true;
+    private Buffer buffer = EMPTY_BUFFER;
+
+    public JsonResponseStreamer(final ServerWebSocket response) {
+      this.response = response;
+    }
+
+    @Override
+    public void write(final int b) throws IOException {
+      singleByteBuf[0] = (byte) b;
+      write(singleByteBuf, 0, 1);
+    }
+
+    @Override
+    public void write(final byte[] bbuf, final int off, final int len) throws IOException {
+      if (buffer != EMPTY_BUFFER) {
+        writeFrame(buffer, false);
+      }
+
+      Buffer buf = Buffer.buffer(len);
+      buf.appendBytes(bbuf, off, len);
+      buffer = buf;
+    }
+
+    private void writeFrame(final Buffer buf, final boolean isFinal) throws IOException {
+      if (response.writeQueueFull()) {
+        LOG.debug("HttpResponse write queue is full pausing streaming");
+        response.drainHandler(e -> paused.release());
+        try {
+          paused.acquire();
+          LOG.debug("HttpResponse write queue is not accepting more data, resuming streaming");
+        } catch (InterruptedException ex) {
+          Thread.currentThread().interrupt();
+          throw new IOException(
+              "Interrupted while waiting for HttpServerResponse to drain the write queue", ex);
+        }
+      }
+
+      if (firstFrame) {
+        response.writeFrame(WebSocketFrame.textFrame(buf.toString(), isFinal));
```

Comment:
Does this work with binary frames as well?

Reply:
To send text messages with Vert.x the first frame must have a string as payload, otherwise it is considered a binary message, for this it necessary to do this test on the first frame.
This way we keep backward compatibility and we do not need to think about the encoding, since websocket text messages are UTF-8 encoded.